src/replication.c

```diff
@@ -2071,14 +2368,164 @@ handle_update_spec_stats(spec_t *spec, zvol_io_hdr_t *hdr, void *resp)
+ */
+static int
+wait_for_ongoing_ios_on_replica(spec_t *spec, replica_t *replica, int sec) {
+	uint64_t no_of_ios_left, current_io_cnt;
+	int retry_count_for_ongoing_ios;
+	int ret = -1;
+	struct timespec last, now, diff;
+	bool io_found = false;	/* Write or Sync IOs */
+
+	ISTGT_LOG("Waiting for IO's to flush on replica(%s:%lu)\n",
+	    replica->replica_id, replica->zvol_guid);
+	// To check whether IO's are flushing to replica or not
+	retry_count_for_ongoing_ios = 5;
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+	ASSERT(spec->quiesce == 1);
+	no_of_ios_left = replica->replica_inflight_write_io_cnt;
+
+	clock_gettime(CLOCK_MONOTONIC_COARSE, &last);
+	timesdiff(CLOCK_MONOTONIC_COARSE, last, now, diff);
+
+	while (diff.tv_sec < sec) {
+		io_found = false;
+		if (replica->replica_inflight_write_io_cnt != 0 ||
+		    replica->replica_inflight_sync_io_cnt != 0) {
+			current_io_cnt = replica->replica_inflight_write_io_cnt;
+			io_found = true;
+		}
+		if (!io_found) {
+			ret = 1;
+			break;
+		}
+		if ((no_of_ios_left - current_io_cnt) > 10) {
+			no_of_ios_left = current_io_cnt;
+			retry_count_for_ongoing_ios = 5;
+		} else if(current_io_cnt != 0) {
+			retry_count_for_ongoing_ios--;
+		}
+
+		if (retry_count_for_ongoing_ios == 0) {
+			ret = -1;
+			break;
+		}
+
+		MTX_UNLOCK(&spec->rq_mtx);
+		/*
+		 * inflight write/sync IOs in spec, or in replica,
+		 * so, wait for some time
+		 */
+		sleep (3);
+		MTX_LOCK(&spec->rq_mtx);
+		timesdiff(CLOCK_MONOTONIC_COARSE, last, now, diff);
+	}
+
+	return ret;
+}
+
+/*
+ * Update_replication_factor updates the replication factor by one,
+ * consistency factor accordingly (n/2 + 1) and known replcia list.
+ * This operation must be performed by taking spec lock and pause IOs
+ * until timeout or operation completes.
+ */
+static int
+update_replication_factor(spec_t *spec, replica_t *replica) {
+	int known_replica_count, data_len;
+	char *data;
+	const char *json_string;
+	struct json_object *jobj;
+	int rf, cf, rc = -1;
+	bool old_replica;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	known_replica_count = get_known_replica_count(spec);
+	old_replica = is_replica_exist(spec, replica->replica_id);
+
+	/* If is_replica_exist return true then it is a kind of replica replace
+	 * or replica movement scenarios. If it return false then it is a replica
+	 * scale up case then to add replica into known replica list current known 
+	 * replica count should be less than desired replication factor.
+	 */
+	if (!old_replica && known_replica_count >= spec->desired_replication_factor) {
+		ISTGT_ERRLOG("failed to update replication/consistency factor "
+		    "known replica count %d is greater than or equal to "
+		    "desired replciation factor %d\n",
+		    known_replica_count, spec->desired_replication_factor);
+		return rc;
+	}
+
+	ASSERT(spec->quiesce == 1);
+	MTX_LOCK(&replica->r_mtx);
+
+	/* Update RF, CF and known replica list*/
+	rf = spec->replication_factor + 1;
```

Comment:
this should not be updated for old_replicas..

Reply:
It is handled in lastest commit